libcudacxx/include/cuda/std/__cuda/semaphore.h

```diff
@@ -38,8 +52,152 @@ class counting_semaphore : public _CUDA_VSTD::__semaphore_base<__least_max_value
   counting_semaphore& operator=(const counting_semaphore&) = delete;
 };
 
+template <thread_scope _Sco, ptrdiff_t __least_max_value = INT_MAX>
+struct fair_counting_semaphore
```

Comment:
Important: We are moving towards modularizing our codebase more and more, so that individual features are isolated from each other. Please move all implementations of a `__fair_binary_semaphore` into a separate header `cuda/std/__semaphore/fair_binary_semaphore.h`